src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+        v8::External::New(isolate, reinterpret_cast<void*>(setter)));
+  }
+
+  if (data) {
+    cbdata->SetInternalField(
+        v8impl::kDataIndex,
+        v8::External::New(isolate, reinterpret_cast<void*>(data)));
+  }
+
+  return cbdata;
+}
+
+}  // end of namespace v8impl
+
+// Intercepts the Node-V8 module registration callback. Converts parameters
+// to NAPI equivalents and then calls the registration callback specified
+// by the NAPI module.
+void napi_module_register_cb(v8::Local<v8::Object> exports,
+                             v8::Local<v8::Value> module,
+                             v8::Local<v8::Context> context,
+                             void* priv) {
+  napi_module* mod = static_cast<napi_module*>(priv);
+  mod->nm_register_func(
+    v8impl::JsEnvFromV8Isolate(context->GetIsolate()),
+    v8impl::JsValueFromV8LocalValue(exports),
+    v8impl::JsValueFromV8LocalValue(module),
+    mod->nm_priv);
+}
+
+#ifndef EXTERNAL_NAPI
+namespace node {
+  // Indicates whether NAPI was enabled/disabled via the node command-line.
+  extern bool load_napi_modules;
+}
+#endif  // EXTERNAL_NAPI
+
+// Registers a NAPI module.
+void napi_module_register(napi_module* mod) {
+  // NAPI modules always work with the current node version.
+  int moduleVersion = NODE_MODULE_VERSION;
+
+#ifndef EXTERNAL_NAPI
+  if (!node::load_napi_modules) {
+    // NAPI is disabled, so set the module version to -1 to cause the module
+    // to be unloaded.
+    moduleVersion = -1;
+  }
+#endif  // EXTERNAL_NAPI
+
+  node::node_module* nm = new node::node_module {
+    moduleVersion,
+    mod->nm_flags,
+    nullptr,
+    mod->nm_filename,
+    nullptr,
+    napi_module_register_cb,
+    mod->nm_modname,
+    mod,  // priv
+    nullptr,
+  };
+  node::node_module_register(nm);
+}
+
+#define RETURN_STATUS_IF_FALSE(condition, status)                       \
+  do {                                                                  \
+    if (!(condition)) {                                                 \
+      return napi_set_last_error((status));                             \
+    }                                                                   \
+  } while (0)
+
+#define CHECK_ARG(arg) RETURN_STATUS_IF_FALSE((arg), napi_invalid_arg)
+
+#define CHECK_MAYBE_EMPTY(maybe, status) \
+  RETURN_STATUS_IF_FALSE(!((maybe).IsEmpty()), (status))
+
+#define CHECK_MAYBE_NOTHING(maybe, status) \
+  RETURN_STATUS_IF_FALSE(!((maybe).IsNothing()), (status))
+
+// NAPI_PREAMBLE is not wrapped in do..while: tryCatch must have function scope.
+#define NAPI_PREAMBLE(e)                                              \
+  CHECK_ARG(e);                                                       \
+  RETURN_STATUS_IF_FALSE(v8impl::TryCatch::lastException().IsEmpty(), \
+                         napi_pending_exception);                     \
+  napi_clear_last_error();                                            \
+  v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)))
```

Comment:
We have made an effort to avoid using ```NAPI_PREAMBLE``` for functions which don't execute JavaScript (see ```napi_is_error``` for example) but we haven't done that exhaustively. I think we took the more conservative approach of including TryCatch when we weren't sure if it was strictly required or not because the guidance seems to be that using TryCatch was cheap enough that this would be fine. If we're sure no JavaScript is getting executed, removing the ```NAPI_PREAMBLE``` seems like the right thing to do.